### ISSUE TITLE
[podofo] Fails to protect PDF file

### ISSUE THREAD

**Issue by derreisende77** [microsoft/vcpkg#24117]
**Describe the bug**
PoDoFo up to now was capable (for several months) of protecting/encrypting PDF files programmatically. Today I performed `vcpkg update` and received various updates for podofo dependencies and my application now crashes while protecting the pdf with podofo. As the problem is related to crypto I think that the latest `openssl` updates may be causing the problem with podofo.
The relevant podofo code that triggers the issue is:

```c++
doc.SetEncrypted("",
                             "passwordstring",
                             PoDoFo::PdfEncrypt::ePdfPermissions_Print |
                             PoDoFo::PdfEncrypt::ePdfPermissions_Accessible |
                             PoDoFo::PdfEncrypt::ePdfPermissions_HighPrint,
                             PoDoFo::PdfEncrypt::ePdfEncryptAlgorithm_AESV2,
                             PoDoFo::PdfEncrypt::ePdfKeyLength_128);
```

Disabling this instruction disables pdf protection and podofo successfully creates a pdf again.
The bug occurs both on windows and macOS.

EDIT: Updated the PoC and the resulting error message.

**Environment**

- OS: Windows 10 21H2 and macOS 12.3.1 Apple Silicon 
- Compiler: Apple Xcode 13.3.1 clang; Visual Studio 2019 latest update

**To Reproduce**

1. ./vcpkg install podofo[fontconfig]

2. Run test code:

   ```c++
   #include <iostream>
   #include <podofo/podofo.h>
   
   int main() {
       try {
           PoDoFo::PdfMemDocument doc;
           // you must provide orig.pdf
           doc.Load("orig.pdf");
           // enabling .SetEncrypted will cause the exception
           doc.SetEncrypted("",
                            "ownerPassword",
                            PoDoFo::PdfEncrypt::ePdfPermissions_Print |
                            PoDoFo::PdfEncrypt::ePdfPermissions_Accessible |
                            PoDoFo::PdfEncrypt::ePdfPermissions_HighPrint,
                            PoDoFo::PdfEncrypt::ePdfEncryptAlgorithm_AESV2,
                            PoDoFo::PdfEncrypt::ePdfKeyLength_128);
           //will be created when no error
           doc.Write("protected.pdf");
       }
       catch (PoDoFo::PdfError &e) {
           e.PrintErrorMsg();
       }
   
       return 0;
   }
   ```

   

3. No compilation errors.

**Expected behavior**
A protected PDF should be created.

**Failure logs**
```
PoDoFo encountered an error. Error: 8 ePdfError_InternalLogic
	Error Description: An internal error occurred.
	Callstack:
	#0 Error Source: /Users/xxx/vcpkg/buildtrees/podofo/src/0e699739c2-e7dcab1c3c.clean/src/podofo/base/PdfEncrypt.cpp:985
		Information: Error initializing RC4 encryption engine
```

**Additional context**
The problem occurs both on Windows and on macOS.

**Comment by derreisende77** [comment]
The error points towards openssl:
```c++
int status = EVP_EncryptInit_ex(rc4, EVP_rc4(), NULL, NULL, NULL);
    if(status != 1)
        PODOFO_RAISE_ERROR_INFO( ePdfError_InternalLogic, "Error initializing RC4 encryption engine" );
```

**Comment by Thomas1664** [comment]
@derreisende77 Can you try to add the following pre-processor definition: `PODOFO_HAVE_OPENSSL_NO_RC4`?

**Comment by derreisende77** [comment]
I couldn´t get it to work by modifying the `CMakeLists.txt` but instead modified `portfile.cmake`:
```
vcpkg_configure_cmake(
    SOURCE_PATH ${SOURCE_PATH}
    PREFER_NINJA
    OPTIONS
        -DPODOFO_BUILD_LIB_ONLY=1
        -DPODOFO_BUILD_SHARED=${PODOFO_BUILD_SHARED}
        -DPODOFO_BUILD_STATIC=${PODOFO_BUILD_STATIC}
        -DPODOFO_NO_FONTMANAGER=${PODOFO_NO_FONTMANAGER}
        -DCMAKE_DISABLE_FIND_PACKAGE_FONTCONFIG=${PODOFO_NO_FONTMANAGER}
        -DCMAKE_DISABLE_FIND_PACKAGE_LIBCRYPTO=${IS_WIN32}
        -DCMAKE_DISABLE_FIND_PACKAGE_LIBIDN=ON
        -DCMAKE_DISABLE_FIND_PACKAGE_CppUnit=ON
        -DCMAKE_DISABLE_FIND_PACKAGE_Boost=ON
        -DPODOFO_HAVE_OPENSSL_NO_RC4=ON
)
```
The PoC now does not throw an exception anymore **BUT it does NOT work as expected**. It is still broken. Generated PDFs opened with Acrobat Reader ask for a password which is always incorrect even if all password are specified as empty in the code. PoDoFo and/or the PDF spec need RC4 legacy to function properly.

**Comment by Adela0814** [comment]
@derreisende77   could you please provide your `CMakeLists.txt` in VS.

**Comment by derreisende77** [comment]
```
cmake_minimum_required(VERSION 3.21)
project(pdf_test)

find_package(Fontconfig REQUIRED)
find_package(Freetype REQUIRED)
find_package(OpenSSL REQUIRED)
find_package(JPEG REQUIRED)
find_package(expat CONFIG REQUIRED)

add_executable(pdf_test main.cpp)

target_link_libraries(pdf_test PRIVATE /Users/XXX/vcpkg/installed/arm64-osx/lib/libpodofo.a)
target_link_libraries(pdf_test PRIVATE OpenSSL::Crypto)
target_link_libraries(pdf_test PRIVATE ${FREETYPE_LIBRARIES})
target_link_libraries(pdf_test PRIVATE ${Fontconfig_LIBRARIES})
target_link_libraries(pdf_test PRIVATE ${JPEG_LIBRARIES})
target_link_libraries(pdf_test PRIVATE /Users/XXX/vcpkg/installed/arm64-osx/lib/libunistring.a)
target_link_libraries(pdf_test PRIVATE expat::expat)
```

I am using CLion both on windows and macOS and CMAKE_TOOLCHAIN_FILE is set in the project.

**Comment by Adela0814** [comment]
Unfortunately, I haven't been able to reproduce this problem. Has it been solved now？@derreisende77

**Comment by derreisende77** [comment]
No. I opened a ticket with the library creators. And I am going to open a ticket for weak cypher support option for openssl.

**Comment by derreisende77** [comment]
BTW they released version 0.9.8 of PoDoFo. Also now available on github as they are migrating to it.
